### DISCUSSION TITLE
Pluggable Token Handling

### DISCUSSION THREAD

**Discussion by esatterwhite** [panva/node-oidc-provider#1256] [Ideas]
Particularly in the case of opaque tokens, It would be great to be able to include the token management as a part of the adapter or some other swappable mechanism. We have a system for managing tokens, and we don't wish to change that. In utilizing this library, it doesn't seem there is a way to define a new token type or control how one is generated.

Ideally, this library should allow implementers to define how a token format is defined and managed.

**Comment by panva** [comment]
Please extend your thought considerably, as is there isn't enough to go off of. As per https://github.com/panva/node-oidc-provider/discussions/new?category=ideas 

> Only fully-fledged ideas with description, examples and clearly stated "why" or "what for" stand a chance of being discussed, otherwise they will be closed and eventually removed.

**Reply by esatterwhite** [reply]
We have a system in place for generating opaque tokens associated with users, or applications - complete with roles, permissions, grants, etc. They are actively in use. Currently the process of getting a token is dated, You have to log into our UI and manually provision. We would like it enable OAuth / OpenIDC workflows as to allow our platform to be a valid identity provider and grant access tokens to people/applications in a standardized way. 

It looks like there is a way to tap into the session and how they are stored using the model adapter, But not the management of the tokens being generated. Enabling opaque tokens assumes that it is acceptable for the frame work to always use nano id. 

Ultimately, I want to be in control of what that token is, and represents

```javascript
class Adapter {
  createToken({scopes, expiration, ...options})
  destroyToken(id)
  validateToken(id)
  refreshToken(refresh_token)
}
```

I may be missing something in the code or docs, but if we want to leverage this library + opaque tokens, we would be unable to leverage our existing AuthN systems.

**Reply by panva** [reply]
Hi @esatterwhite 

There are means in place to put custom claims in issued access tokens, those then being available either via introspection or directly in the token if that token is a structured JWT token.

You would use your existing authnz system to enrich the tokens issued by the server, not replace them.

**Reply by esatterwhite** [reply]
Technically, If there is a way for me to define how the tokens are created, then they are issued by the server. 
But the format of our tokens has meaning, is can be validated, Can be identified by Security scanners / SAST tools, and are associated to more in our system that what can reasonably be expressed with claims. nanoids simply aren't sufficient for what we want to do. We also don't want diverging formats depending on how you came to obtain them.

My goal isn't necessarily to replace the tokens the after the server creates, I want to define how they are created.

Especially in the case of opaque tokens, the spec doesn't define a format, only that they be unique, In that respect, It shouldn't matter to this frame work what does that or how it is done. Its opaque to the consumer, That doesn't mean it must also be opaque to the provider.

Help me to understand the resistance to the idea in general

**Reply by esatterwhite** [reply]
Access tokens specifically, not id tokens, refresh tokens, etc.  just to clarify

**Reply by panva** [reply]
> Help me to understand the resistance to the idea in general

Don't interpret my suggestions to use that which is already there as resistance.

> Can be identified by Security scanners / SAST tools

I think that's a neat idea there to be able to define an opaque token prefix such as `op_at_${value}` for Access Tokens, Refresh Tokens, Authorization Codes, Device Codes, CIBA Requests. At the same time I'd think of a MAC digest of the prefix and value as part of the issued opaque value.

**Reply by esatterwhite** [reply]
> At the same time I'd think of a MAC digest of the prefix and value as part of the issued opaque value.

We have a bit of both. We have an identifiable prefix of different token types, we HMAC either a random value, or some user supplied key, and include a crc33 hash. Then keep a checksum so the public tokens are never actually persisted. Which is important for various compliance reasons at our company.
This, in addition to a tiered and mailable acl system. 

There are quite a number of reasons to be in control of the management of the tokens themselves.

**Reply by esatterwhite** [reply]
What else would be needed for this to become a concrete feature?

**Reply by weyert** [reply]
Looks like a nice idea!

**Reply by esatterwhite** [reply]
@panva just checking back here. What would it take to push this forward?

**Reply by panva** [reply]
@esatterwhite I'm open to having the issued opaque tokens and cookie values have a CRC checksum and a fixed prefix. That alone is a feature I would get behind. Having this "pluggable" [like so](https://github.com/panva/node-oidc-provider/discussions/1256#discussioncomment-9014065) I am not interested in pursuing.
